### ISSUE TITLE
Deserializing a URI requires parsing it

### ISSUE THREAD

**Issue by jdm** [servo/servo#20412]
https://github.com/servo/rust-url/blob/04d5ec1e7f3e70c850ff2e7ba7d83d446926defe/url_serde/src/lib.rs#L180-L188 means that any time we send a Url over an IPC channel we have to re-parse it when receiving it. This seems inefficient.

**Comment by SimonSapin** [comment]
The correct solution is to support serde 1.0 inside the url crate so that the impls can access private fields. However this probably requires removing support for previous serde versions, which is a breaking change that will need to be in url 2.0. This in turn requires https://github.com/servo/servo/issues/19825 since hyper 0.10 depends on url 1.x. And if we’re doing url 2.0, we’ll probably want to take that opportunity to also make a number of other desired breaking changes that have been accumulating since 1.0.0.

**Comment by nox** [comment]
I don't see how moving serde support directly in the url crate fixes that issue, given that still is driven by serde and anyone would be able to, say, write some hand-crafted JSON to create an invalid `Url` value.

**Comment by nox** [comment]
The real fix is to stop using Serde for IPC, and rework our entire IPC infrastructure, but that's a massive investment of developer time.

**Comment by SimonSapin** [comment]
Couldn’t a hand-crafted non-serde IPC message create an invalid `Url` just the same?

**Comment by nox** [comment]
Yes definitely, but we would control it. I don’t think serde support itself in the url crate should have such a hatch in safe code.

-- 
Anthony Ramine
